sdk/identity/src/token_credentials/client_certificate_credentials.rs

```diff
@@ -0,0 +1,231 @@
+    fn default() -> Self {
+        Self {
+            authority_host: authority_hosts::AZURE_PUBLIC_CLOUD.to_owned(),
+            send_certificate_chain: false,
+        }
+    }
+}
+
+impl CertificateCredentialOptions {
+    /// Create a new TokenCredentialsOptions. default() may also be used.
+    pub fn new(authority_host: String, send_certificate_chain: bool) -> Self {
+        Self {
+            authority_host,
+            send_certificate_chain,
+        }
+    }
+    /// Set the authority host for authentication requests.
+    pub fn set_authority_host(&mut self, authority_host: String) {
+        self.authority_host = authority_host
+    }
+
+    /// The authority host to use for authentication requests.  The default is
+    /// https://login.microsoftonline.com.
+    pub fn authority_host(&self) -> &str {
+        &self.authority_host
+    }
+
+    /// Enable/disable sending the certificate chain
+    pub fn set_send_certificate_chain(&mut self, send_certificate_chain: bool) {
+        self.send_certificate_chain = send_certificate_chain
+    }
+
+    /// Whether certificate chain is sent as part of the request or not. Default is
+    /// set to true
+    pub fn send_certificate_chain(&self) -> bool {
+        self.send_certificate_chain
+    }
+}
+
+/// Enables authentication to Azure Active Directory using a client certificate that
+/// was generated for an App Registration.
+///
+/// In order to use subject name validation send_cert_chain option must be set to true
+/// The certificate is expected to be in base64 encoded der
+pub struct ClientCertificateCredential {
+    tenant_id: String,
+    client_id: String,
+    client_certificate: String,
+    client_certificate_pass: String,
+    options: CertificateCredentialOptions,
+}
+
+impl ClientCertificateCredential {
+    /// Create a new ClientCertificateCredential
+    pub fn new(
+        tenant_id: String,
+        client_id: String,
+        client_certificate: String,
+        client_certificate_pass: String,
+        options: CertificateCredentialOptions,
+    ) -> ClientCertificateCredential {
+        ClientCertificateCredential {
+            tenant_id,
+            client_id,
+            client_certificate,
+            client_certificate_pass,
+            options,
+        }
+    }
+
+    fn options(&self) -> &CertificateCredentialOptions {
+        &self.options
+    }
+
+    fn sign(jwt: &str, pkey: &PKey<Private>) -> Result<Vec<u8>, ErrorStack> {
+        let mut signer = Signer::new(MessageDigest::sha256(), pkey)?;
+        signer.update(jwt.as_bytes())?;
+        signer.sign_to_vec()
+    }
+
+    fn get_thumbprint(cert: &X509) -> Result<DigestBytes, ErrorStack> {
+        let der = cert.to_der()?;
+        let digest = hash(MessageDigest::sha1(), &der)?;
+        Ok(digest)
+    }
+
+    fn as_jwt_part(part: &[u8]) -> String {
+        base64::encode(&part)
+            .replace('+', "-")
+            .replace('/', "_")
+            .replace('=', "")
```

Comment:
For consistency with other crates, use `base64::*_config(..., BASE64_URL_SAFE)`.

Reply:
i was not aware of this part. Fixed.